pkg/virt-controller/watch/drain/evacuation/evacuation.go

```diff
@@ -371,18 +371,27 @@ func (c *EvacuationController) sync(node *k8sv1.Node, vmisOnNode []*virtv1.Virtu
 		Effect: k8sv1.TaintEffectNoSchedule,
 	}
 
+	numOfActiveMigrations := len(activeMigrations)
+
 	vmisToMigrate := vmisToMigrate(node, vmisOnNode, taint)
 	if len(vmisToMigrate) == 0 {
 		return nil
 	}
 
 	migrationCandidates, nonMigrateable := c.filterRunningNonMigratingVMIs(vmisToMigrate, activeMigrations)
-	activeMigrationsFromThisSourceNode := c.numOfVMIMForThisSourceNode(vmisOnNode, activeMigrations)
+	numOfActiveMigrationsFromThisSourceNode := c.numOfVMIMForThisSourceNode(vmisOnNode, activeMigrations)
 	maxParallelMigrationsPerOutboundNode :=
 		int(*c.clusterConfig.GetMigrationConfiguration().ParallelOutboundMigrationsPerNode)
 	maxParallelMigrations := int(*c.clusterConfig.GetMigrationConfiguration().ParallelMigrationsPerCluster)
-	freeSpotsPerCluster := maxParallelMigrations - len(activeMigrations)
-	freeSpotsPerThisSourceNode := maxParallelMigrationsPerOutboundNode - activeMigrationsFromThisSourceNode
+
+	if (numOfActiveMigrations >= maxParallelMigrations) ||
+		(numOfActiveMigrationsFromThisSourceNode >= maxParallelMigrationsPerOutboundNode) {
```

Comment:
I might be missing something here, but I'm not sure I understand how this can happen by performing the reproduction steps on the BZ.

Please keep me honest here, but IIUC this can only happen if someone/something else (other than the evacuation controller) posts migration objects. Is this correct?

Reply:
@iholder101 Yes this can be also a way to reproduce it, but not only. Another way, maybe more complex, is that I've described in the BZ steps, when you are using `nodeSelector` in the VMI. You label only one worker, when you drain it all the migrations are created but stuck because no other node exists with same label, then you reduce the `maxParallelMigrations` and you cause a negative `freeSpots`

Reply:
thanks for the clarification!